src: remote: fix error message thrown when `kw ssh` fails using a remote config file

davidbtadokoro opened the pull request:
This patch corrects the error message when a failed `kw ssh` occurs using a remote config file instead of explicitly typing the IP, user and port of the host (using `kw ssh --remote | -r`). It also adds a test to verify if the correct error message is being thrown to the user in three distinct cases:
- Explicitly typing the IP, user and port;
- Using a remote config file;
- When no remote config file is found (without explicitly declaring the host infos).

Closes #695

Commit:
tests: remote: add test for ssh_connection_failure_message function

Although this function does not influence in the logic of the module 'kw
ssh', it displays useful information for the user in case of a failed
connection (mainly the IP, user and port attempted). As such, a test to
check if the correct information is being display in some failure cases
is important.

WIP #695

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: remote: fix error message from failed ssh connection

When the user tries to connect via 'kw ssh' using a remote config file
instead of explicitly typing the IP, user and port (with 'kw ssh -r')
and fails, those three informations are not displayed to the user. This
was probably due to the change from adding this explicit declaration of
the host, because the 'ssh_connection_failure_message' function expects
that those informations are already set in the 'remote_parameters'
variable. In the case of using a remote config file, these are not set
and we need to retrieve them from the file to avoid them being displayed
as blank. Also added a message if no remote config file is found.

Closes #695

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented:
Hi @davidbtadokoro ,
 Could you rebase your change in the latest code from the unstable branch?

davidbtadokoro commented:
Thanks for noticing @rodrigosiqueira !
Doing it now!

davidbtadokoro commented:
Noticed that there are no tests for the `is_ssh_connection_configured()` function.

Also, when no remote config file is found (without explicitly defining the host with `kw ssh -r`) `kw ssh` tries anyway to establish a connection, resulting in a `Bad port '@'` thrown to the user.

Me and @pomba123 are creating tests for this function and fixing this issue of trying to connect anyway when we already know that the connection is going to fail.

Commit:
tests: remote: add tests for is_ssh_connection_configured function

This important function, as it is called for every 'kw ssh' call, did
not have a test. As such, we included a unit test for it considering the
three cases of an attempted connection: 1) explicitly stating the host
with 'kw ssh -r'; 2) using a remote config file; 3) using a remote
config file, but kw could not find one.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>
Co-authored-by: Everaldo Gomes Junior <everaldogjr@gmail.com>

Commit:
src: remote: handle remote config file not found case

In the 'is_ssh_connection_configured' function there was no error
handling when kw did not find a remote config file. With this commit,
rather than ignoring this case and trying to establish a connection
anyway we return an error code of no entry (ENOENT) and avoid extra
computation when we know that it will fail.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>
Co-authored-by: Everaldo Gomes Junior <everaldogjr@gmail.com>

kwy95 commented on src/remote.sh:
```diff
@@ -28,6 +28,8 @@ function is_ssh_connection_configured()
     if [[ -n "${remote_file}" ]]; then
       ssh_cmd="ssh -q -o StrictHostKeyChecking=accept-new -o BatchMode=yes -o ConnectTimeout=5 -F ${remote_file} ${remote_file_host} exit"
+    else
+      return 2 # ENOENT
```
Again, you implemented a test that expects this behavior before this behavior was implemented. As I said, you shouldn't introduce broken tests on a commit.

kwy95 reviewed (changes requested):
The biggest takeaway is that you should keep your commits as atomic as possible, firstly by ensuring that all the changes in the commit are necessary to achieve the commits purpose, and nothig else. But also, ensure that all tests and functionality works at every commit, so that the tool could be used at any point of it's history.

davidbtadokoro commented:
Many thanks for the feedback @kwy95 !!

It makes sense, the tool should work at any point. I actually made the fixes before the tests, but idk why I thought it was best to commit the tests before 😭

Also, thanks for pointing the other improvements.
I think I am going to close this PR and make 2 others (of one commit each to correct the failure message bug and the error handling) to clean the messy git history.
